Fix r29/gosp swap ; Add register save/restore to tests that need it

androm3da opened the pull request.

quic-mathbern commented on tests/test_mmu.S:
```diff
@@ -105,6 +109,11 @@ GUEST_event_vectors:
+	memd(r29) = r1:0
+	memd(r29+#8) = r3:2
+	memd(r29+#16) = r5:4
+
```
Don't we need to save p3:0 as well?

androm3da replied:
Seems like we did in `GUEST_event_trap0()`.  And we are overwriting `p0`.

So, yeah, maybe that makes sense.

androm3da replied:
@taylorsimpson do we need to save the predicate registers here?

androm3da replied:
@taylorsimpson fixed this, updated with the changes.  Thanks @quic-mathbern for finding it.

taylorsimpson committed:
Add register save/restore to tests that need it

QEMU VM trace will warn when guest event handlers don't save/restore
registers.

Signed-off-by: Taylor Simpson <ltaylorsimpson@gmail.com>

taylorsimpson committed:
Fix r29/gosp swap

The patch to use guest registers instead of memory locations replaced
    memd(r24+#CONTEXT_g32) = r13:12
with
    gbadva = r13
This leaves out the assignment of g2(gosp) = r12

The correct assignment is
    g3:2 = r13:12

Signed-off-by: Taylor Simpson <ltaylorsimpson@gmail.com>

quic-mathbern commented on tests/test_mmu.S:
```diff
@@ -41,6 +43,7 @@ _start:
 	r25 = pc
 	r24 = ##(guest_pt-_start)
 	r24 = add(r24, r25)
+	p0 = cmp.eq(r25, #0)
```
I don't quite understand this line, what is it for?

quic-mathbern replied:
Taylor explained this offline:

> It's a way to trigger a warning in the QEMU VM trace when the predicate registers aren't saved/restored.  The original value of p0 (must be set by minivm bootup code) is 0xff, and the test event handler sets it to 1.  Therefore, no change is detected.  We change it to zero here, so the change would be visible to the trace.

I think we could add a small comment before it for the future.

androm3da replied:
Ok, added.
